akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress
+
+import akka.actor.{ ActorRef, ActorSystem }
+import akka.event.LoggingAdapter
+
+final class InboundActorRefCompressionTable(
+  system:                   ActorSystem,
+  heavyHitters:             TopHeavyHitters[ActorRef],
+  onNewHeavyHitterDetected: (ActorRef, Int) ⇒ Unit)
+  extends InboundCompressionTable[ActorRef](system.log, heavyHitters, _.path.toSerializationFormat, onNewHeavyHitterDetected) {
+
+  override def decompress(idx: Int): ActorRef =
+    if (idx == 0) system.deadLetters
+    else super.decompress(idx)
+}
+
+/**
+ * Handles counting and detecting of heavy-hitters and compressing them via a table lookup.
+ * Mutable and not thread-safe.
+ *
+ * Compression flow goes like:
+ * [1]  on each message we add the actor path here
+ * [2]  if it becomes a heavy hitter, we allocate an identifier for it and invoke the callback
+ * [3]> the callback for example then triggers an CompressionAdvertisement to the receiving side
+ */
+// TODO would like the function to be named
+// TODO should the onHeavyHitter be inside HeavyHitters?
+class InboundCompressionTable[T](
+  log:                      LoggingAdapter,
+  heavyHitters:             TopHeavyHitters[T],
+  convertKeyToString:       T ⇒ String,
+  onNewHeavyHitterDetected: (T, Int) ⇒ Unit) {
+  require(heavyHitters != null, "heavyHitters must not be null")
+
+  // TODO calibrate properly
+  private[this] val cms = new CountMinSketch(100, 100, 0)
+
+  private[this] var compressionAllocations = Map.empty[Int, T] // TODO replace with a specialized LongMap
+  private[this] var currentCompressionId = InboundCompressionTable.CompressionAllocationCounterStart
+
+  /**
+   * Decompress given identifier into original String representation.
+   *
+   * @throws UnknownCompressedIdException if given id is not known, this may indicate a bug – such situation should not happen.
+   */
+  def decompress(idx: Int): T = {
+    log.debug(s"  --- decompress [$idx] => ${compressionAllocations.get(idx)}")
```

Comment:
done, and also hidden under additional compression.debug flag, it's only ever useful when debugging compression - not poluting general debug output